app/src/main/java/app/myzel394/alibi/effect/opengl/renderpass/WatermarkRenderPass.kt

```diff
@@ -0,0 +1,120 @@
+import app.myzel394.alibi.effect.opengl.core.GlProgram
+import app.myzel394.alibi.effect.opengl.core.Gles20Wrapper
+import app.myzel394.alibi.effect.opengl.core.createTexture
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+
+private const val TEXT_SIZE_MULTIPLY_VIDEO_SCREEN_RATIO = 30
+private const val TAG = "WatermarkRenderPass"
+
+/**
+ * Used to render a watermark.
+ */
+class WatermarkRenderPass(
+    private val context: Context,
+    private val vertexShader: DefaultVertexShader,
+    private val fragmentShader: FragmentTextureShader,
+    private val program: GlProgram,
+    private val gles20Wrapper: Gles20Wrapper = Gles20Wrapper.DEFAULT,
+) : RenderPass {
+
+    constructor(
+        context: Context,
+        vertexShader: DefaultVertexShader =
+            DefaultVertexShader(),
+        fragmentShader: FragmentTextureShader =
+            FragmentTextureShader(ShaderSource.FRAGMENT_TEXTURE_2D),
+    ) : this(context, vertexShader, fragmentShader, GlProgram(vertexShader, fragmentShader))
+
+    private var texId: Int = GL_INVALID
+
+    private val textFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SS", Locale.US);
+
+    private val paint = Paint().apply {
+        setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD));
+        isAntiAlias = true
+        color = Color.WHITE
+    }
+
+    private var bitmap: Bitmap? = null
+
+    private var canvas: Canvas? = null
+
+    override fun init() {
+        vertexShader.init()
+        fragmentShader.init()
+        program.init()
+
+        program.use()
+        vertexShader.loadLocations(program.programId)
+        vertexShader.configureAttributes()
+        texId = gles20Wrapper.createTexture(GLES20.GL_TEXTURE_2D)
+    }
+
+    override fun draw(texMatrix: FloatArray, mvpMatrix: FloatArray, surfaceSize: Size) {
+        program.use()
+
+        gles20Wrapper.glEnable(GLES20.GL_BLEND)
+        gles20Wrapper.glBlendFunc(GLES20.GL_ONE, GLES20.GL_ONE_MINUS_SRC_ALPHA)
+
+        gles20Wrapper.glActiveTexture(GLES20.GL_TEXTURE0)
+        gles20Wrapper.glBindTexture(fragmentShader.textureType, texId)
+
+        val bitmap = drawTextToBitmap(surfaceSize)
+
+        gles20Wrapper.glViewport(0, 0, bitmap.width, bitmap.height)
+        GLUtils.texImage2D(fragmentShader.textureType, 0, bitmap, 0)
+
+        vertexShader.configureMatrix(texMatrix, mvpMatrix)
+        gles20Wrapper.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, 4)
+
+        gles20Wrapper.glDisable(GLES20.GL_BLEND)
+    }
+
+    private fun drawTextToBitmap(surfaceSize: Size): Bitmap {
+        val text = textFormat.format(Date())
+
+        val bitmap = this.bitmap ?: let {
+            val videoToScreenRatio = surfaceSize.height * 1f / context.resources.displayMetrics.heightPixels
+            paint.textSize = videoToScreenRatio * TEXT_SIZE_MULTIPLY_VIDEO_SCREEN_RATIO
+            paint.setShadowLayer(videoToScreenRatio * 1f, 0f, 0f, Color.BLACK)
```

Comment:
Note:
- Bot video resolution and screen size affect the watermark experience, as most use cases play the video right on the device.
- For a consistent viewing experience, watermark size is calculated based on the ratio of Video Resolution / Screen Size